File: frontend/lib/models/toy.dart
// lib/models/toy.dart
import 'user.dart';

class Toy {
  final String id;
  final String name;
  final String description;
  final String imageUrl;
  final String location;
  final double latitude;
  final double longitude;
  final String owner;
  final double? price;
  final bool? isAvailable;
  bool isFavorite = false;
  final Map<String, dynamic>? coordinates;

  Toy({
    required this.id,
    required this.name,
    required this.description,
    required this.imageUrl,
    required this.location,
    required this.latitude,
    required this.longitude,
    required this.owner,
    this.price,
    this.isAvailable,
    this.coordinates,
  });

  Toy copyWith() {
    return Toy(
      id: id,
      name: name,
      description: description,
      imageUrl: imageUrl,
      location: location,
      latitude: latitude,
      longitude: longitude,
      owner: owner,
      price: price,
      isAvailable: isAvailable,
      coordinates: coordinates,
    );
  }

  factory Toy.fromJson(Map<String, dynamic> json) {
    final ownerValue = json['owner'];
    String ownerId = '';
    if (ownerValue is Map<String, dynamic>) {
      ownerId = ownerValue['_id'] ?? '';
    } else if (ownerValue is String) {
      ownerId = ownerValue;
    }

    return Toy(
      id: json['_id'] ?? '',
      name: json['name'] ?? '',
      description: json['description'] ?? '',
      imageUrl: json['imageUrl'] ?? '',
      location: json['location'] ?? '',
      latitude: (json['latitude'] as num).toDouble(),
      longitude: (json['longitude'] as num).toDouble(),
      owner: ownerId,
      price: json['price']?.toDouble(),
      isAvailable: json['isAvailable'],
      coordinates: json['coordinates'] is Map<String, dynamic>
          ? Map<String, dynamic>.from(json['coordinates'])
          : null,
    );
  }

  Map<String, dynamic> toJson() {
    return {
      '_id': id,
      'name': name,
      'description': description,
      'imageUrl': imageUrl,
      'location': location,
    };
  }
}

File: backend/models/Toy.js
// models/Toy.js
const mongoose = require("mongoose");
const User = require("./User");

const toySchema = new mongoose.Schema(
  {
    name: {
      type: String,
      required: true,
    },
    description: {
      type: String,
      required: true,
    },
    imageUrl: {
      type: String,
      required: true,
    },
    location: {
      type: String,
      required: true,
    },
    latitude: {
      type: Number,
      required: true,
    },
    longitude: {
      type: Number,
      required: true,
    },
    coordinates: {
      type: {
        type: String,
        enum: ["Point"],
        default: "Point",
      },
      coordinates: {
        type: [Number], // [longitude, latitude]
        required: true,
      },
    },
    owner: {
      type: mongoose.Schema.Types.ObjectId,
      ref: "User",
      required: true,
    },
  },
  {
    timestamps: true,
  }
);

// Ajouter un index géospatial
toySchema.index({ coordinates: "2dsphere" });

// Déplacer les hooks AVANT la création du modèle
toySchema.post("save", async function (doc, next) {
  try {
    console.log("🔄 Déclenchement hook post-save pour jouet:", doc._id);
    console.log(
      "🔍 Owner ID:",
      doc.owner,
      "Type:",
      typeof doc.owner,
      "Valid:",
      mongoose.isValidObjectId(doc.owner)
    );

    const result = await mongoose
      .model("User")
      .findByIdAndUpdate(
        doc.owner,
        { $inc: { totalListings: 1 } },
        { new: true, runValidators: true }
      );

    console.log("✅ Résultat mise à jour utilisateur:", result);
    next();
  } catch (error) {
    console.error("❌ Erreur hook post-save:", error);
    next(error);
  }
});

// Modifier le hook existant pour les suppressions
toySchema.post(["deleteOne", "findOneAndDelete"], async function (doc) {
  await User.findByIdAndUpdate(
    doc.owner,
    { $inc: { totalListings: -1 } },
    { new: true }
  );
});

const Toy = mongoose.model("Toy", toySchema);

module.exports = Toy;
